nixos/modules/services/misc/pghero.nix

```diff
@@ -0,0 +1,138 @@
+    enable = lib.mkEnableOption (lib.mdDoc "PgHero service");
+    package = lib.mkPackageOptionMD pkgs "pghero" { };
+
+    listenAddress = lib.mkOption {
+      type = lib.types.str;
+      example = "[::1]:3000";
+      description = lib.mdDoc ''
+        `hostname:port` to listen for HTTP traffic.
+
+        This is bound using the systemd socket activation.
+      '';
+    };
+
+    extraArgs = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+      default = [ ];
+      description = lib.mdDoc ''
+        Additional command-line arguments for the systemd service.
+      '';
+    };
+
+    settings = lib.mkOption {
+      type = settingsFormat.type;
+      default = { };
+      example = {
+        databases = {
+          primary = {
+            url = "<%= ENV['PRIMARY_DATABASE_URL'] %>";
+          };
+        };
+      };
+      description = lib.mdDoc ''
+        PgHero configuration. Refer to the [PgHero documentation] for more
+        details.
+
+        [PgHero documentation]: https://github.com/ankane/pghero/blob/master/guides/Linux.md#multiple-databases
+      '';
+    };
+
+    environment = lib.mkOption {
+      type = lib.types.attrsOf lib.types.str;
+      default = { };
+      description = lib.mdDoc ''
+        Environment variables to set for the service. Secrets should be
+        specified using {option}`environmentFile`.
+      '';
+    };
+
+    environmentFiles = lib.mkOption {
+      type = lib.types.listOf lib.types.path;
+      default = [ ];
+      description = lib.mdDoc ''
+        File to load environment variables from. Loaded variables override
+        values set in {option}`environment`.
+      '';
+    };
+
+    extraGroups = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+      default = [ ];
+      example = [ "tlskeys" ];
+      description = lib.mdDoc ''
+        Additional groups for the systemd service.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.sockets.pghero = {
+      unitConfig.Description = "PgHero HTTP socket";
+      wantedBy = [ "sockets.target" ];
+      listenStreams = [ cfg.listenAddress ];
+    };
+
+    systemd.services.pghero = {
+      description = "PgHero performance dashboard for PostgreSQL";
+      wantedBy = [ "multi-user.target" ];
+      requires = [ "pghero.socket" ];
+      after = [ "pghero.socket" "network.target" ];
+
+      environment = {
+        RAILS_ENV = "production";
+        PGHERO_CONFIG_PATH = settingsFile;
+      } // cfg.environment;
+
+      serviceConfig = {
+        Type = "notify";
+        WatchdogSec = "10";
+
+        ExecStart = utils.escapeSystemdExecArgs ([
+          (lib.getExe cfg.package)
+          "--bind-to-activated-sockets"
+          "only"
+        ] ++ cfg.extraArgs);
+        Restart = "always";
+
+        WorkingDirectory = "${cfg.package}/share/pghero";
+
+        EnvironmentFile = cfg.environmentFiles;
+        SupplementaryGroups = cfg.extraGroups;
+
+        DynamicUser = true;
```

Comment:
DynamicUser probably works poorly with postgres socket authentication.

Reply:
Do you mean [peer authentication](https://www.postgresql.org/docs/current/auth-peer.html)? That shouldn’t be an issue since systemd integrates with glibc NSS. For Unix socket permissions there is extraGroups option that sets `SupplementaryGroups=`.

Reply:
Yes but that can be really tricky to use with such authentication types. At the point postgres starts, the user/group for pghero don't exist yet and I am not sure how postgres handles that. From my experience with other services doing socket auth, that does not always works great.

Reply:
> At the point postgres starts, the user/group for pghero don't exist yet and I am not sure how postgres handles that.

Same as if PgHero is deployed after PostgreSQL? I’m not really sure what is the issue you are trying to highlight here.

At work, we are using exactly this configuration for production servers (DynamicUser that connects via local peer/ident authentication). NixOS test also covers this case (I’ve added the comment to document that).

Reply:
> Same as if PgHero is deployed after PostgreSQL?

It depends on the start ordering of the services and how postgres handles users. If you say it works, then I take your word for that.